src/tofnd/client.rs

```diff
@@ -0,0 +1,264 @@
+#![allow(dead_code)]
+
+use std::convert::TryInto;
+
+use async_trait::async_trait;
+use error_stack::{FutureExt, IntoReport, Report, Result, ResultExt};
+use mockall::automock;
+use tonic::transport::Channel;
+use tonic::{Response, Status};
+
+use super::proto::{
+    key_presence_response::Response as KeyPresenceEnum, keygen_response::KeygenResponse, multisig_client,
+    sign_response::SignResponse, KeyPresenceRequest, KeygenRequest, SignRequest,
+};
+use super::{error::Error, error::TofndError, PublicKey, Signature};
+
+#[automock]
+#[async_trait]
+pub trait MultisigClient {
+    async fn keygen(&mut self, request: KeygenRequest) -> Result<KeygenResponse, Status>;
+    async fn sign(&mut self, request: SignRequest) -> Result<SignResponse, Status>;
+    async fn key_presence(&mut self, request: KeyPresenceRequest) -> Result<KeyPresenceEnum, Status>;
+}
+
+#[async_trait]
+impl MultisigClient for multisig_client::MultisigClient<Channel> {
+    async fn keygen(&mut self, request: KeygenRequest) -> Result<KeygenResponse, Status> {
+        self.keygen(request)
+            .await
+            .and_then(|response| {
+                response
+                    .into_inner()
+                    .keygen_response
+                    .ok_or_else(|| Status::internal("keygen response is empty"))
+            })
+            .into_report()
+    }
+
+    async fn sign(&mut self, request: SignRequest) -> Result<SignResponse, Status> {
+        self.sign(request)
+            .await
+            .and_then(|response| {
+                response
+                    .into_inner()
+                    .sign_response
+                    .ok_or_else(|| Status::internal("sign response is empty"))
+            })
+            .into_report()
+    }
+
+    async fn key_presence(&mut self, request: KeyPresenceRequest) -> Result<KeyPresenceEnum, Status> {
+        self.key_presence(request)
+            .await
+            .map(Response::into_inner)
+            .and_then(|response| {
+                KeyPresenceEnum::from_i32(response.response)
+                    .ok_or_else(|| Status::internal("invalid key presence response"))
+            })
+            .into_report()
+    }
+}
+
+pub struct Client<T: MultisigClient> {
+    client: T,
+    party_uid: String,
+}
+
+impl<T: MultisigClient> Client<T> {
+    pub fn new(client: T, party_uid: String) -> Self {
+        Self { client, party_uid }
+    }
+
+    pub async fn keygen(&mut self, key_uid: String) -> Result<PublicKey, Error> {
+        let req = KeygenRequest {
+            key_uid,
+            party_uid: self.party_uid.clone(),
+        };
+
+        self.client
+            .keygen(req)
+            .change_context(Error::Grpc)
+            .await
+            .and_then(|response| match response {
+                KeygenResponse::PubKey(pub_key) => pub_key.try_into().map_err(|val| {
+                    Report::from(Error::KeygenFailed).attach_printable(format!("{{ invalid value = {:?} }}", val))
```

Comment:
```suggestion
                    Report::from(Error::KeygenFailed).attach_printable(format!("{{ invalid_value = {:?} }}", val))
```
since it's json I don't think space will fly here